### DISCUSSION TITLE
Multiline docstrings and more

### DISCUSSION THREAD

**Discussion by jeffmikels** [vlang/v#14611] [Ideas]
The current implementation of docstrings has three insufficiencies:

- The current docstring parser does not allow for multi-line comments. Prepending `//` at the start of the line is acceptable for three or four lines at a time, but it creates inconvenience to the author of the code, and discourages detailed documentation.
- The current parser only allows for single-line "Example" code. This is insufficient for documenting functions such as `sort_with_compare`. Consider the official documentation on that function to see the need for multi-line examples.
- The current docstring parser does not automatically handle links to other keywords. The parser should be aware of all public symbols in the current module at least so as to generate links to related documentation.
- Although some modules have what appear to be docstrings at the top of the module, these docstrings are not actually included in any generated documentation.

As a result, the official documentation is not automatically generated, nor can it be, and therefore, the documentation in the code is maintained independently from the official language documentation. Furthermore, we are required to use `README.md` files to add extensive documentation to our modules or to the `vlib` modules.

Possible solution:

I'm a fan of the way the Dart language handles docstrings. The following are the significant components that make writing documented code in Dart easy and make docstrings more useful for the end user:

- Dart uses `///` a triple slash to indicate a docstring comment.
- The Dart VS Code extension recognizes this and automatically prepends `///` at the beginning of the line while also preserving indentation so that code can easily be written inside a docstring.
- By convention, inside a docstring square brackets are used to indicate a symbol that the parser can treat as a reference to other code in the current file or elsewhere.
- There is no special treatment of "Example" in a dart docstring. The entire docstring is parsed as a multi-line markdown document, so embedded code is specified with backticks either in block form or inline.
- The Dart language server and VS Code extension parses all docstrings in all local files and included modules to display documentation on hover.

**Comment by JalonSolov** [comment]
Using `///` has been discussed on discord, and was already decided as something V wants.  It simply hasn't been done, yet.

As for square brackets, the problem is how to distinguish between what should be a link, and what is just an example of code - Not an official `/// Example:` line, just something like "/// To access the first element of an array, use arr[0]".  Probably won't be too successful turning that into a link.

My suggestion is to use something that is _not_ part of the V syntax, such as # instead of square brackets.  So if you wanted a link to `foo` you would use `#foo#` in the comment.

**Comment by ghost** [comment]
Do we need more documentation than we have now in the V code for readability?
_`README.md` should include design principles, detailed usage, examples more_, and the current `//` and `/**/` should be sufficient in the code, not changing the philosophy that the code itself is the documentation.
Trying to include more documentation in a `.v` file seems to fundamentally shake up the design of V and will contribute to damaging the readability of V.
Is Dart a language that puts readability first in the first place?
`README.md` should be more complete.

**Comment by jeffmikels** [comment]
The distinction between what is code and what is an automatically determined link is made with the use of backticks. in your example:

```v
/// To access the first element of an array, use `arr[0]`
```

In other words, the `///` will indicate "We're in a docstring" that will be parsed using markdown with one extension that words in square brackets will be automatically matched to module keywords/symbols, but square brackets will not be parsed that way if they are contained in a code block.

**Comment by JalonSolov** [comment]
That's... somewhat arcane, and will rely on documentation describing the various ways to do things.

The V philosophy is "one way" to do things (which, I know, it hasn't always stuck to, but it tries).

There should be "one way" to specify links, and it should always apply whether you're in a docstring or not (although, of course, it won't have much effect if it's not).  In other words, one shouldn't be forced to use \`\` around code, nor have to be careful about using [] in docstrings.

**Comment by larpon** [comment]
If we support markdown already, then `[link](https://url.net)` is the "one way".

**Comment by jeffmikels** [comment]
Perhaps an example will help you understand what I'm going for:

Consider the webpage here: https://modules.vlang.io/sokol.audio.html#setup

( by the way, did you notice that Github just violated the original markdown spec by automatically generating that link? )

I digress...

If you look at the documentation for the `sokol.audio.setup` function, you will see that it requires an object called `C.saudio_desc`. Now, on a computer screen, one can easily look at the table of contents on the right column to see the link at the bottom for where that struct is described, but on a mobile device, the table of contents in the right column is removed entirely. One must scroll through the whole document to find the location of `C.saudio_desc` and there is no visible indication that such a struct is even documented on this page.

Additionally, once you find the definition of `C.saudio_desc` at the bottom of the documentation, you will discover new  types like `FNStreamingCB` and `FnStreamingCBWithUserData` again, with no indication that they are documented on this page.

One solution to this problem is to make the document generator replace all potential symbols (types, structs, functions) throughout the document with links to the page where documentation for that symbol may be found. Perhaps that's the right solution in the long run.

Another solution is to require the developer of a module to manually include cross-reference links in their docstrings. This is the method I already employed to get the docstring links working in the documentation for `sokol.audio.FnStreamingCBWithUserData` https://modules.vlang.io/sokol.audio.html#FnStreamingCBWithUserData

You can see the link is there, but I implemented it with this docstring in the code:

```v
// This function operates the same way as [[FNStreamingCB](#FNStreamingCB)] but it passes …
```

Not only was this cumbersome to write, it also assumes a knowledge of the docstring html output that will create an anchor named `#FNStreamingCB`. Additionally, it formatted it in the final html in a way that is not pleasant. It created a standard link in a standard font, but it would have looked better as inline code with a link.

Additionally, adding that link in the code makes the code comment LESS readable for anyone who is not browsing the documentation in the html format.

Therefore, the following is the best compromise, giving V docstrings better readability in the code and in the generated documentation, and preserving in V standards, "one-way-to-do-docstrings"

```v
/// This function operates the same way as [FNStreamingCB] but it passes …
```

Adding a third slash is the one-way to tell V we are in a docstring.

Putting a bracket around a linkable keyword is the one-way to flag an item in the docstring as V symbol that should result in an internal link if possible.

Extended Markdown is the one-way to do everything else... including when creating example code blocks (i.e. triple-backticks)

By adopting this convention for docstrings, the docstring generator can be updated to create the links automatically in the html renderer, while leaving them alone in the other renderers.

**Comment by larpon** [comment]
Thanks for the explanation

**Comment by JalonSolov** [comment]
How about a compromise...  `[#FNStreamingCB]` would mean a link, while `[FNStreamingCB]` would just be code?

It covers both things we're discussing... it is simpler/shorter to type, yet is also distinct from regular code, where `#` is an invalid character.

**Comment by jeffmikels** [comment]
That's an interesting suggestion, and a workable compromise. I'm not convinced that brackets inside a docstring would "look like code" to anyone, but the Hash symbol does hint that it's going to be turned into a link. For people familiar with Markdown, it might be confusing, but if the convention is widely employed in the source, I'm sure they'd figure it out soon enough.

Nevertheless, conventions are things that need buy-in. Perhaps we should wait until Alex has a chance to chime in.
